File: loadercomplex-forge-12/src/main/java/com/enderzombi102/loadercomplex/forge12/impl/entity/ForgePlayer.java
package com.enderzombi102.loadercomplex.forge12.impl.entity;

import com.enderzombi102.loadercomplex.api.entity.Player;
import com.enderzombi102.loadercomplex.api.utils.Gamemode;
import com.enderzombi102.loadercomplex.api.utils.Position;
import com.enderzombi102.loadercomplex.forge12.impl.utils.BlockUtils;
import net.minecraft.client.Minecraft;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.text.TextComponentString;
import net.minecraft.world.GameType;

public class ForgePlayer extends ForgeLivingEntity implements Player {
	private final EntityPlayer wrappedEntity;

	public ForgePlayer(EntityPlayer player) {
		super(player);
		this.wrappedEntity = player;
	}

	@Override
	public boolean isSleeping() {
		return this.wrappedEntity.isPlayerSleeping();
	}

	@Override
	public Position getBedLocation() {
		return BlockUtils.toPosition( this.wrappedEntity.getBedLocation() );
	}

	@Override
	public int getScore() {
		return this.wrappedEntity.getScore();
	}

	@Override
	public void addScore(int score) {
		this.wrappedEntity.addScore(score);
	}

	@Override
	public void sendMessage(String msg) {
		this.wrappedEntity.sendMessage( new TextComponentString( msg ) );
	}

	@Override
	public void setRespawnPoint(Position pos) {
		this.wrappedEntity.setSpawnPoint( new BlockPos( pos.x, pos.y, pos.z ), true );
	}

	@Override
	public int getFoodLevel() {
		return this.wrappedEntity.getFoodStats().getFoodLevel();
	}

	@Override
	public void setFoodLevel(int level) {
		this.wrappedEntity.getFoodStats().setFoodLevel( level );
	}

	@Override
	public float getSaturationLevel() {
		return this.wrappedEntity.getFoodStats().getSaturationLevel();
	}

	@Override
	public void setSaturationLevel(float level) {
		this.wrappedEntity.getFoodStats().setFoodSaturationLevel( level );
	}

	@Override
	public Gamemode getGamemode() {
		GameType type;
		if ( this.wrappedEntity.getEntityWorld().isRemote )
			type = Minecraft.getMinecraft().playerController.getCurrentGameType();
		else
			type = ( (EntityPlayerMP) this.wrappedEntity ).interactionManager.getGameType();
		if ( type == GameType.NOT_SET )
			throw new IllegalStateException("Player has no gamemode!");
		return Gamemode.valueOf( type.getName().toUpperCase() );
	}

	@Override
	public int getExperienceLevel() {
		return this.wrappedEntity.experienceLevel;
	}

	@Override
	public boolean isPlayer() {
		return true;
	}

	@Override
	public Player asPlayer() {
		return this;
	}
}

File: loadercomplex-forge/src/ver122/kotlin/com/enderzombi102/loadercomplex/forge122/impl/entity/ForgePlayer.kt
package com.enderzombi102.loadercomplex.forge122.impl.entity

import com.enderzombi102.loadercomplex.api.math.Vec3i
import com.enderzombi102.loadercomplex.api.minecraft.entity.PlayerEntity
import com.enderzombi102.loadercomplex.api.minecraft.util.Gamemode
import com.enderzombi102.loadercomplex.forge122.impl.utils.toLC
import net.minecraft.client.Minecraft
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.util.math.BlockPos
import net.minecraft.util.text.TextComponentString
import net.minecraft.world.GameType
import java.util.*

class ForgePlayer(private val wrappedEntity: EntityPlayer) : ForgeLivingEntity(
	wrappedEntity
),
	PlayerEntity {
	override fun isSleeping(): Boolean {
		return wrappedEntity.isPlayerSleeping
	}

	override fun getBedLocation(): Optional<Vec3i> {
		return Optional.ofNullable(wrappedEntity.getBedLocation()).map(BlockPos::toLC)
	}

	override fun getScore(): Int {
		return wrappedEntity.score
	}

	override fun addScore(score: Int) {
		wrappedEntity.addScore(score)
	}

	override fun sendMessage(msg: String) {
		wrappedEntity.sendMessage(TextComponentString(msg))
	}

	override fun setRespawnPoint(pos: Vec3i) {
		wrappedEntity.setSpawnPoint(BlockPos(pos.x, pos.y, pos.z), true)
	}

	override fun getFoodLevel(): Int {
		return wrappedEntity.foodStats.foodLevel
	}

	override fun setFoodLevel(level: Int) {
		wrappedEntity.foodStats.foodLevel = level
	}

	override fun getSaturationLevel(): Float {
		return wrappedEntity.foodStats.saturationLevel
	}

	override fun setSaturationLevel(level: Float) {
		wrappedEntity.foodStats.setFoodSaturationLevel(level)
	}

	override fun getGamemode(): Gamemode {
		val type = if (wrappedEntity.entityWorld.isRemote) {
			Minecraft.getMinecraft().playerController.currentGameType
		} else {
			(wrappedEntity as EntityPlayerMP).interactionManager.gameType
		}
		check(type != GameType.NOT_SET) { "Player has no gamemode!" }
		return Gamemode.valueOf(type.getName().uppercase(Locale.getDefault()))
	}

	override fun getExperienceLevel(): Int {
		return wrappedEntity.experienceLevel
	}

	override fun isPlayer(): Boolean {
		return true
	}

	override fun asPlayer(): PlayerEntity {
		return this
	}
}
